core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java

```diff
@@ -96,17 +97,20 @@ public void iterateStatus(AlluxioURI path, ListStatusPOptions options,
       // Otherwise, needs more complicated logic inside the cache,
       // that might not worth the effort of caching.
       super.iterateStatus(path, options, action);
+      return;
     }
 
-    List<URIStatus> statuses = mMetadataCache.listStatus(path);
-    if (statuses == null) {
+    List<URIStatus> cachedStatuses = mMetadataCache.listStatus(path);
+    if (cachedStatuses == null) {
+      List<URIStatus> statuses = new ArrayList<>();
       super.iterateStatus(path, options, status -> {
-        mMetadataCache.put(path, status);
```

Comment:
`mMetadataCache.put(path, status);` is not caching `ls` result of a dir and it should be `status.getPath`, not `path`